fhir-remote-index/src/main/java/com/ibm/fhir/remote/index/sharded/ShardedPostgresMessageHandler.java

```diff
@@ -0,0 +1,1088 @@
+import java.sql.SQLException;
+import java.sql.Types;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+import com.ibm.fhir.database.utils.api.IDatabaseTranslator;
+import com.ibm.fhir.database.utils.common.ResultSetReader;
+import com.ibm.fhir.database.utils.postgres.PostgresTranslator;
+import com.ibm.fhir.persistence.exception.FHIRPersistenceException;
+import com.ibm.fhir.persistence.index.DateParameter;
+import com.ibm.fhir.persistence.index.LocationParameter;
+import com.ibm.fhir.persistence.index.NumberParameter;
+import com.ibm.fhir.persistence.index.ProfileParameter;
+import com.ibm.fhir.persistence.index.QuantityParameter;
+import com.ibm.fhir.persistence.index.ReferenceParameter;
+import com.ibm.fhir.persistence.index.RemoteIndexMessage;
+import com.ibm.fhir.persistence.index.SearchParameterValue;
+import com.ibm.fhir.persistence.index.SecurityParameter;
+import com.ibm.fhir.persistence.index.StringParameter;
+import com.ibm.fhir.persistence.index.TagParameter;
+import com.ibm.fhir.persistence.index.TokenParameter;
+import com.ibm.fhir.remote.index.api.BatchParameterValue;
+import com.ibm.fhir.remote.index.api.IdentityCache;
+import com.ibm.fhir.remote.index.batch.BatchDateParameter;
+import com.ibm.fhir.remote.index.batch.BatchLocationParameter;
+import com.ibm.fhir.remote.index.batch.BatchNumberParameter;
+import com.ibm.fhir.remote.index.batch.BatchProfileParameter;
+import com.ibm.fhir.remote.index.batch.BatchQuantityParameter;
+import com.ibm.fhir.remote.index.batch.BatchReferenceParameter;
+import com.ibm.fhir.remote.index.batch.BatchSecurityParameter;
+import com.ibm.fhir.remote.index.batch.BatchStringParameter;
+import com.ibm.fhir.remote.index.batch.BatchTagParameter;
+import com.ibm.fhir.remote.index.batch.BatchTokenParameter;
+import com.ibm.fhir.remote.index.database.BaseMessageHandler;
+import com.ibm.fhir.remote.index.database.CodeSystemValue;
+import com.ibm.fhir.remote.index.database.CommonCanonicalValue;
+import com.ibm.fhir.remote.index.database.CommonCanonicalValueKey;
+import com.ibm.fhir.remote.index.database.CommonTokenValue;
+import com.ibm.fhir.remote.index.database.CommonTokenValueKey;
+import com.ibm.fhir.remote.index.database.LogicalResourceIdentKey;
+import com.ibm.fhir.remote.index.database.LogicalResourceIdentValue;
+import com.ibm.fhir.remote.index.database.LogicalResourceValue;
+import com.ibm.fhir.remote.index.database.ParameterNameValue;
+import com.ibm.fhir.remote.index.database.PlainMessageHandler;
+import com.ibm.fhir.remote.index.database.PreparedStatementWrapper;
+
+/**
+ * Loads search parameter values into the target FHIR schema on
+ * a PostgreSQL database.
+ * TODO refactor to try and share more processing with the {@link PlainMessageHandler}
```

Comment:
is it worth a separate technical debt issue?

Reply:
#3711